File: LeetCode/Medium/34 - Find First and Last Position of Element in Sorted Array/34 - Find First and Last Position of Element in Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 var searchRange = function(nums, target) {
    const n = nums.length;
    
    const start = binarySearch(0, n - 1, nums, target);
    const end = binarySearch(0, n - 1, nums, target + 1) - 1;

    return (nums[start] == target ? [start, end] : [-1, -1]);
};

var binarySearch = function(lo, hi, ar, target) {
    while (lo <= hi) {
        const mid = (lo + hi) >> 1;
        
        if (ar[mid] < target)
            lo = mid + 1;
        else
            hi = mid - 1;
    }
    
    return lo;
};

File: LeetCode/Medium/34 - Find First and Last Position of Element in Sorted Array/34 - Find First and Last Position of Element in Sorted Array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int s = binarySearch(nums, target);
        int e = binarySearch(nums, target + 1) - 1;
        int n = nums.size();
        if (s < 0 || e < 0 || s >= n || e >= n || nums[s] != target) return {-1, -1};
        return {s, e};
    }
    
    int binarySearch(vector<int> &vc, int target) {
        int lo = 0, hi = vc.size() - 1, mid;
        while (lo <= hi) {
            mid = lo + ((hi - lo) >> 1);
            
            if (vc[mid] < target)
                lo = mid + 1;
            else
                hi = mid - 1;
        }
        return lo;
    }
};